plugin/pkg/scheduler/volumebinder/volume_binder.go

```diff
@@ -0,0 +1,80 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package volumebinder
+
+import (
+	"time"
+
+	"k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+	utilfeature "k8s.io/apiserver/pkg/util/feature"
+	coreinformers "k8s.io/client-go/informers/core/v1"
+	storageinformers "k8s.io/client-go/informers/storage/v1"
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/kubernetes/pkg/controller/volume/persistentvolume"
+	"k8s.io/kubernetes/pkg/features"
+)
+
+// VolumeBinder sets up the volume binding library and manages
+// the volume binding operations with a queue.
+type VolumeBinder struct {
+	Binder    persistentvolume.SchedulerVolumeBinder
+	BindQueue *workqueue.Type
+}
+
+// NewVolumeBinder sets up the volume binding library and binding queue
+func NewVolumeBinder(
+	client clientset.Interface,
+	pvcInformer coreinformers.PersistentVolumeClaimInformer,
+	pvInformer coreinformers.PersistentVolumeInformer,
+	nodeInformer coreinformers.NodeInformer,
+	storageClassInformer storageinformers.StorageClassInformer) *VolumeBinder {
+
+	return &VolumeBinder{
+		Binder:    persistentvolume.NewVolumeBinder(client, pvcInformer, pvInformer, nodeInformer, storageClassInformer),
+		BindQueue: workqueue.NewNamed("podsToBind"),
+	}
+}
+
+// NewFakeVolumeBinder sets up a fake volume binder and binding queue
+func NewFakeVolumeBinder(config *persistentvolume.FakeVolumeBinderConfig) *VolumeBinder {
+	return &VolumeBinder{
+		Binder:    persistentvolume.NewFakeVolumeBinder(config),
+		BindQueue: workqueue.NewNamed("podsToBind"),
+	}
+}
+
+// Run starts a goroutine to handle the binding queue with the given function.
+func (b *VolumeBinder) Run(bindWorkFunc func(), stopCh <-chan struct{}) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
+		go wait.Until(bindWorkFunc, time.Second, stopCh)
+
+		<-stopCh
+		b.BindQueue.ShutDown()
+	}
+}
+
+// DeletePodBindings will delete the cached volume bindings for the given pod.
+func (b *VolumeBinder) DeletePodBindings(pod *v1.Pod) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.VolumeScheduling) {
```

Comment:
similar to above, probably we don't need the check.

Reply:
DeletePodBindings will grab/release a lock, which shouldn't really see much contention, but who knows?  I guess it's only called outside the main scheduler thread (via informer events or error handler).

Reply:
The alternative is to put the check on the call site (if there are not many places where this is called). That would make things a bit more clear IMO.

Reply:
I added all the calls in this [commit](https://github.com/kubernetes/kubernetes/pull/55039/commits/47ea0839b65d655b10f7c9ae916801d0458a8be4).   There are 6 callers.